cluster-autoscaler/cloudprovider/clusterapi/internal/clusterapi_machineset.go

```diff
@@ -0,0 +1,150 @@
+	}
+	glog.Infof("machineset: %q has %d replicas", m.MachineSet.Name, *m.MachineSet.Spec.Replicas)
+	return int(*m.MachineSet.Spec.Replicas)
+}
+
+func (m *clusterMachineSet) SetSize(nreplicas int) error {
+	ms, err := m.clusterapi.MachineSets(m.MachineSet.Namespace).Get(m.MachineSet.Name, v1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("Unable to get machineset %q: %v", m.MachineSet.Name, err)
+	}
+
+	newMachineSet := ms.DeepCopy()
+	replicas := int32(nreplicas)
+	newMachineSet.Spec.Replicas = &replicas
+
+	_, err = m.clusterapi.MachineSets(m.MachineSet.Namespace).Update(newMachineSet)
+	if err != nil {
+		return fmt.Errorf("Unable to update number of replicas of machineset %q: %v", m.MachineSet.Name, err)
+	}
+
+	return nil
+}
+
+func (m *clusterMachineSet) Nodes() ([]string, error) {
+	return m.nodes, nil
+}
+
+func (m *clusterMachineSet) DeleteNodes(nodenames []string) error {
+	if len(nodenames) == 0 {
+		return nil
+	}
+
+	snapshot := m.getClusterState()
+
+	for _, nodename := range nodenames {
+		name, exists := snapshot.NodeMap[machineSetID(m.MachineSet)][nodename]
+		if !exists {
+			return fmt.Errorf("cannot map nodename %q to machine", nodename)
+		}
+		machine, err := m.clusterapi.Machines(m.MachineSet.Namespace).Get(name, v1.GetOptions{})
+		if err != nil {
+			return fmt.Errorf("cannot get machine %s/%s: %v", m.MachineSet.Namespace, name, err)
+		}
+
+		machine = machine.DeepCopy()
+
+		if machine.Annotations == nil {
+			machine.Annotations = map[string]string{}
+		}
+
+		// Annotate machine that it is the chosen one.
+		// TODO(frobware) - don't hardcode this name
+		machine.Annotations["sigs.k8s.io/cluster-api-delete-machine"] = time.Now().String()
+
+		_, err = m.clusterapi.Machines(m.MachineSet.Namespace).Update(machine)
+		if err != nil {
+			return fmt.Errorf("unable to update machine %q: %v", machine.Name, err)
+		}
+	}
+
+	replicas := m.Replicas()
+	if replicas-len(nodenames) <= 0 {
+		return fmt.Errorf("unable to delete %d machines in %s, machine replicas are <= 0 ", len(nodenames), m)
+	}
+
+	return m.SetSize(replicas - len(nodenames))
+}
+
+func (m *clusterMachineSet) String() string {
+	return fmt.Sprintf("%s/%s", m.Namespace(), m.Name())
+}
+
+func parseLabel(ms *v1alpha1apis.MachineSet, label string, result *int) {
+	val, exists := ms.Labels[label]
+	if !exists {
+		glog.Infof("machineset %s/%s has no label named %q", ms.Namespace, ms.Name, label)
+		return
+	}
+
+	u, err := strconv.ParseUint(val, 10, 32)
+	if err != nil {
+		glog.Errorf("machineset %s/%s: cannot parse %q as an integral value: %v", ms.Namespace, ms.Name, val, err)
+		return
+	}
+
+	*result = int(u)
+}
+
+func newClusterMachineSet(m *clusterManager, ms *v1alpha1apis.MachineSet, nodes []string) *clusterMachineSet {
+	cms := clusterMachineSet{
+		clusterManager: m,
+		MachineSet:     ms,
+		nodes:          nodes,
+		nodeSet:        make(map[string]bool),
+	}
+
+	parseLabel(ms, "sigs.k8s.io/cluster-api-autoscaler-node-group-min-size", &cms.min)
```

Comment:
please handle errors correctly instead of eating them

Reply:
Indeed. :-/

I went back and forth on this - if this fails to parse do we want to a) log and mark the min/max as zero which would mean the machine set is not "offered" to the autoscaler as something it can use or b) just simply treat this as an error which would mean getting any snapshot of the cluster would fail and the autoscaler would have zero machine sets to work with until corrected.

Reply:
I did the latter. If there's an error then `getClusterSnapshot()` errors. No new machinesets detected or updated until the error is corrected.